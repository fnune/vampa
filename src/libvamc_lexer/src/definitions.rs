/// Base of the numeric literal.
#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub enum Base {
    /// No prefix.
    Decimal,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub enum LiteralKind {
    /// "12"
    Int { base: Base },
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub enum TokenKind {
    /// An inline comment:
    /// "# This is an inline comment"
    InlineComment,
    /// A block comment:
    /// "#[ This is a block comment #]"
    /// Nested comments are terminated if they all have balanced opening "#[" and closing "#]" signs.
    BlockComment { terminated: bool },
    /// "12"
    Literal { kind: LiteralKind },
    /// A keyword or a user-defined identifier.
    Identifier,
    /// ":"
    Colon,
    /// ";"
    Semicolon,
    /// Any sequence of whitespace characters.
    Whitespace,
    /// "="
    EqualitySign,
    /// "+"
    PlusSign,
    /// "-"
    MinusSign,
    /// "/"
    Slash,
    /// "%"
    PercentSign,
    /// "*"
    Star,
    /// "{"
    OpeningBrace,
    /// "}"
    ClosingBrace,
    /// A token not expected by the lexer.
    Unknown,
}

#[derive(Debug, PartialEq)]
pub struct Token {
    pub kind: TokenKind,
    pub value: String,
}
