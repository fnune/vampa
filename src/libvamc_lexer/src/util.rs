pub fn is_identifier_start(character: char) -> bool {
    unicode_xid::UnicodeXID::is_xid_start(character)
}

pub fn is_identifier_continuation(character: char) -> bool {
    unicode_xid::UnicodeXID::is_xid_continue(character)
}

// https://github.com/rust-lang/rust/blob/20c50444650f90f266ab2a46afd5089b4c01a28c/src/librustc_lexer/src/lib.rs#L222
pub fn is_whitespace(character: char) -> bool {
    match character {
    // Usual ASCII suspects
    | '\u{0009}' // \t
    | '\u{000A}' // \n
    | '\u{000B}' // vertical tab
    | '\u{000C}' // form feed
    | '\u{000D}' // \r
    | '\u{0020}' // space

    // NEXT LINE from latin1
    | '\u{0085}'

    // Bidi markers
    | '\u{200E}' // LEFT-TO-RIGHT MARK
    | '\u{200F}' // RIGHT-TO-LEFT MARK

    // Dedicated whitespace characters from Unicode
    | '\u{2028}' // LINE SEPARATOR
    | '\u{2029}' // PARAGRAPH SEPARATOR
    => true,
    _ => false,
  }
}

pub fn is_numeric_literal(character: char) -> bool { ('0'..='9').contains(&character) }
