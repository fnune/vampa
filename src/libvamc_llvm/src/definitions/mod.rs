use std::collections::HashMap;

use inkwell::{
    builder::Builder,
    context::Context,
    module::Module,
    values::{FunctionValue, PointerValue},
};
use vamc_errors::Diagnostic;

pub type CompilerResult<T> = Result<T, Diagnostic>;

pub struct Compiler<'a, 'ctx> {
    pub context: &'ctx Context,
    pub builder: &'a Builder<'ctx>,
    pub module: &'a Module<'ctx>,
    pub function_value: Option<&'a FunctionValue<'a>>,

    pub variables: HashMap<String, PointerValue<'ctx>>,
}
