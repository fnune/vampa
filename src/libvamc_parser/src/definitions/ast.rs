type Identifier = String;

#[derive(Debug, PartialEq)]
pub struct Module {
    pub file_name: String,
    pub statements: Vec<Box<Statement>>,
}

#[derive(Debug, PartialEq)]
pub struct Statement {
    pub kind: StatementKind,
}

#[derive(Debug, PartialEq)]
pub enum StatementKind {
    VariableDeclaration(VariableDeclaration),
    FunctionDeclaration(FunctionDeclaration),
    /// An expression statement expected to be implicitly returned.
    Return(Box<Expression>),
}

#[derive(Debug, PartialEq)]
pub struct VariableDeclaration {
    pub name: Box<Identifier>,
    pub typ: Box<Typ>,
    pub value: Box<Expression>,
}

#[derive(Debug, PartialEq)]
pub struct FunctionDeclaration {
    pub name: Box<Identifier>,
    pub parameters: Parameters,
    pub return_typ: Box<Typ>,
    pub body: Box<Expression>,
}

#[derive(Debug, PartialEq)]
pub struct Parameter {
    pub name: Box<Identifier>,
    pub typ: Box<Typ>,
}

pub type Parameters = Vec<Box<Parameter>>;

#[derive(Debug, PartialEq)]
pub struct Expression {
    pub kind: ExpressionKind,
}

#[derive(Debug, PartialEq)]
pub enum ExpressionKind {
    /// A literal expression such as `42`.
    Literal(Literal),
    /// A function call such as `sum 20 10` FunctionCall(F, P) where F is the
    /// function expression itself and P is the Vector of parameters.
    FunctionCall(Box<Expression>, Vec<Box<Expression>>),
    /// A binary operation such as `+ 20 20`
    BinaryOperation(BinaryOperation),
    /// Blocks always evaluate to something, so they're also expressions.
    Block(Box<Block>),
    /// A reference to a variable.
    VariableReference(Box<Identifier>),
}

#[derive(Debug, PartialEq)]
pub struct Block {
    /// Must end with a `Statement` of `StatementKind::Return`.
    pub statements: Vec<Box<Statement>>,
}

/// Even though Vampa uses prefix notation, we still refer
/// to binary operands as <operator> <left> <right>.
#[derive(Debug, PartialEq)]
pub struct BinaryOperation {
    pub kind: BinaryOperationKind,
    pub left: Box<Expression>,
    pub right: Box<Expression>,
}

/// Should eventually support other commonly expected binary operations.
#[derive(Debug, PartialEq)]
pub enum BinaryOperationKind {
    Addition,
}

#[derive(Debug, PartialEq)]
pub struct Literal {
    pub kind: LiteralKind,
}

#[derive(Debug, PartialEq)]
pub enum LiteralKind {
    Int(i128, LiteralIntType),
}

/// Should eventually support signed and unsigned as well.
#[derive(Debug, PartialEq)]
pub enum LiteralIntType {
    Unsuffixed,
}

#[derive(Debug, PartialEq)]
pub struct Typ {
    pub kind: TypKind,
}

#[derive(Debug, PartialEq)]
pub enum TypKind {
    Int(IntType),
    /// A placeholder for a type that must be inferred.
    Infer,
}

impl Typ {
    pub fn infer() -> Typ {
        Typ {
            kind: { TypKind::Infer },
        }
    }
}

/// Should eventually support the types advertised in the README.
#[derive(Debug, PartialEq)]
pub enum IntType {
    I32,
}
