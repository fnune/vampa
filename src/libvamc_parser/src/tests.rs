#[cfg(test)]
mod blocks {
    mod blocks;
}

#[cfg(test)]
mod variables {
    mod variable_declarations;
}

#[cfg(test)]
mod functions {
    mod functions_implicit_return_type;
    mod functions_explicit_return_type;
}
