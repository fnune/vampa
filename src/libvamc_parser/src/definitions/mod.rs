pub mod ast;

use std::{iter::Peekable, vec::IntoIter};

use vamc_errors::Diagnostic;
use vamc_lexer::definitions::*;

pub type ParserResult<T> = Result<T, Diagnostic>;

pub struct Parser {
    /// The tokens from the lexer.
    pub tokens: Peekable<IntoIter<Token>>,
    /// The latest token.
    pub current: Token,
}
