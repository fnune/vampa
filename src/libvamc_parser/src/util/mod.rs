use vamc_lexer::definitions::*;

pub fn is_binary_operator(token: &Token) -> bool {
    matches!(
        token.kind,
        TokenKind::PlusSign
            | TokenKind::MinusSign
            | TokenKind::Slash
            | TokenKind::PercentSign
            | TokenKind::Star
    )
}

pub fn is_parameters_list_termination(token: &Token) -> bool {
    matches!(token.kind, TokenKind::ClosingBrace | TokenKind::Semicolon)
}

pub fn is_identifier_exactly(identifier: &str, token: &Token) -> bool {
    match token.kind {
        TokenKind::Identifier => token.value == identifier,
        _ => false,
    }
}

pub fn is_keyword_and(token: &Token) -> bool {
    is_identifier_exactly("and", token)
}
pub fn is_keyword_apply(token: &Token) -> bool {
    is_identifier_exactly("apply", token)
}
pub fn is_keyword_fun(token: &Token) -> bool {
    is_identifier_exactly("fun", token)
}
pub fn is_keyword_let(token: &Token) -> bool {
    is_identifier_exactly("let", token)
}
pub fn is_keyword_of(token: &Token) -> bool {
    is_identifier_exactly("of", token)
}
pub fn is_keyword_returning(token: &Token) -> bool {
    is_identifier_exactly("returning", token)
}

pub fn is_keyword(token: &Token) -> bool {
    is_keyword_and(token)
        || is_keyword_apply(token)
        || is_keyword_fun(token)
        || is_keyword_let(token)
        || is_keyword_of(token)
        || is_keyword_returning(token)
}
