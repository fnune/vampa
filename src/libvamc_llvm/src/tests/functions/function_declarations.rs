use insta::assert_snapshot;
use vamc_parser::definitions::ast::*;

use crate::tests::util::*;

#[test]
fn typed_function_declaration() {
    let function_declaration = FunctionDeclaration {
        name: Box::new("return_one".to_owned()),
        parameters: vec![Box::new(Parameter {
            name: Box::new("x".to_owned()),
            typ: Box::new(Typ {
                kind: TypKind::Int(IntType::I32),
            }),
        })],
        return_typ: Box::new(Typ {
            kind: TypKind::Int(IntType::I32),
        }),
        body: Box::new(Expression {
            kind: ExpressionKind::Literal(Literal {
                kind: LiteralKind::Int(1, LiteralIntType::Unsuffixed),
            }),
        }),
    };

    with_compiler(|mut compiler, _| {
        compiler
            .compile_function_declaration(function_declaration)
            .unwrap();

        assert_snapshot!(compiler.module.print_to_string().to_string_lossy())
    });
}

#[test]
fn typed_ternary_function_declaration() {
    let function_declaration = FunctionDeclaration {
        name: Box::new("return_one".to_owned()),
        parameters: vec![
            Box::new(Parameter {
                name: Box::new("a".to_owned()),
                typ: Box::new(Typ {
                    kind: TypKind::Int(IntType::I32),
                }),
            }),
            Box::new(Parameter {
                name: Box::new("b".to_owned()),
                typ: Box::new(Typ {
                    kind: TypKind::Int(IntType::I32),
                }),
            }),
            Box::new(Parameter {
                name: Box::new("c".to_owned()),
                typ: Box::new(Typ {
                    kind: TypKind::Int(IntType::I32),
                }),
            }),
        ],
        return_typ: Box::new(Typ {
            kind: TypKind::Int(IntType::I32),
        }),
        body: Box::new(Expression {
            kind: ExpressionKind::Literal(Literal {
                kind: LiteralKind::Int(1, LiteralIntType::Unsuffixed),
            }),
        }),
    };

    with_compiler(|mut compiler, _| {
        compiler
            .compile_function_declaration(function_declaration)
            .unwrap();

        assert_snapshot!(compiler.module.print_to_string().to_string_lossy())
    });
}
