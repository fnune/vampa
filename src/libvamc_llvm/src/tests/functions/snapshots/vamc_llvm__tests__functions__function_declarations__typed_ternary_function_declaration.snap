---
source: src/libvamc_llvm/src/tests/functions/function_declarations.rs
expression: compiler.module.print_to_string().to_string_lossy()
---
; ModuleID = 'test_module'
source_filename = "test_module"

define i32 @test_function() {
test_block:
}

define i32 @return_one(i32 %a, i32 %b, i32 %c) {
return_one:
  ret i32 1
}

